File: src/tools/PdfsToTxt.cpp
#include "PdfsToTxt.h"
#include "../SecurityManager.h"
#include "../Utils.h"
#include <QDesktopServices>
#include <QDirIterator>
#include <QFile>
#include <QFileDialog>
#include <QHBoxLayout>
#include <QLabel>
#include <QMessageBox>
#include <QProcess>
#include <QRegularExpression>
#include <QStandardPaths>
#include <QTextStream>
#include <QUrl>
#include <QVBoxLayout>
#include <algorithm>

namespace GOL {

PdfsToTxt::PdfsToTxt(QWidget *parent) : QDialog(parent) {
  // Security Check
  SecurityManager::instance().checkAndAct();

  setWindowTitle("GOLEVENTS - PDF NAME CLEANER PRO 📄");
  resize(800, 700);
  setStyleSheet(
      QString("background-color: %1; color: white;").arg(Utils::BG_COLOR));

  QVBoxLayout *mainLayout = new QVBoxLayout(this);
  mainLayout->setContentsMargins(40, 30, 40, 40);
  mainLayout->setSpacing(20);

  // Header
  QLabel *titleLabel = new QLabel("📄 PDF NAME CLEANER");
  titleLabel->setStyleSheet(
      QString("font-size: 30px; font-weight: bold; color: %1;")
          .arg(Utils::ACCENT_COLOR));
  mainLayout->addWidget(titleLabel);

  QLabel *subTitle = new QLabel(
      "Extract 'Sector-Row-Seat' only from filenames into a TXT list.");
  subTitle->setStyleSheet("font-size: 13px; color: gray;");
  mainLayout->addWidget(subTitle);

  // Action Frame
  QHBoxLayout *actionLayout = new QHBoxLayout();
  m_pathEdit = new QLineEdit("No folder selected...");
  m_pathEdit->setReadOnly(true);
  m_pathEdit->setFixedHeight(45);
  m_pathEdit->setStyleSheet(
      QString("background-color: #0d0d0d; border: 1px solid %1; border-radius: "
              "8px; padding-left: 10px;")
          .arg(Utils::CARD_BORDER));

  m_btnBrowse = new QPushButton("BROWSE");
  m_btnBrowse->setFixedSize(120, 45);
  m_btnBrowse->setCursor(Qt::PointingHandCursor);
  m_btnBrowse->setStyleSheet(QString("background-color: %1; color: white; "
                                     "border-radius: 8px; font-weight: bold;")
                                 .arg(Utils::ACCENT_COLOR));

  actionLayout->addWidget(m_pathEdit);
  actionLayout->addWidget(m_btnBrowse);
  mainLayout->addLayout(actionLayout);

  // Process Button
  m_btnProcess = new QPushButton("🚀 START EXTRACTION & SAVE");
  m_btnProcess->setFixedHeight(55);
  m_btnProcess->setCursor(Qt::PointingHandCursor);
  m_btnProcess->setStyleSheet(
      "QPushButton { background-color: #10b981; color: black; border-radius: "
      "12px; font-weight: bold; font-size: 16px; }"
      "QPushButton:hover { background-color: #059669; }");
  mainLayout->addWidget(m_btnProcess);

  // Log area
  m_logArea = new QTextEdit();
  m_logArea->setReadOnly(true);
  m_logArea->setFont(QFont("Consolas", 12));
  m_logArea->setStyleSheet(
      QString("background-color: #050505; color: #00FF94; border: 1px solid "
              "%1; border-radius: 10px; padding: 10px;")
          .arg(Utils::CARD_BORDER));
  mainLayout->addWidget(m_logArea);

  // Connections
  connect(m_btnBrowse, &QPushButton::clicked, this, &PdfsToTxt::browseFolder);
  connect(m_btnProcess, &QPushButton::clicked, this, &PdfsToTxt::runCleaner);
}

void PdfsToTxt::browseFolder() {
  QString dir = QFileDialog::getExistingDirectory(this, "Select PDF Folder");
  if (!dir.isEmpty()) {
    m_pathEdit->setText(dir);
  }
}

void PdfsToTxt::log(const QString &msg, bool clear) {
  if (clear)
    m_logArea->clear();
  m_logArea->append("> " + msg);
}

bool PdfsToTxt::naturalSort(const QString &a, const QString &b) {
  // Basic natural sort implementation for "Sector-Row-Seat"
  QStringList partsA = a.split('-');
  QStringList partsB = b.split('-');

  if (partsA.size() < 3 || partsB.size() < 3)
    return a < b;

  // Compare Sector (Alpha-numeric)
  // Logic: if both are numeric, compare as int. Else compare as string.
  // Python logic: if p1 is digit, p1=(0, int), else (1, p1).

  bool isNumA;
  int secNumA = partsA[0].toInt(&isNumA);

  bool isNumB;
  int secNumB = partsB[0].toInt(&isNumB);

  if (isNumA && isNumB) {
    if (secNumA != secNumB)
      return secNumA < secNumB;
  } else if (isNumA != isNumB) {
    // One is number, one is string. Number comes first (0 vs 1 prefix)
    return isNumA;
  } else {
    // Both strings
    if (partsA[0] != partsB[0])
      return partsA[0] < partsB[0];
  }

  // Compare Row (Numeric)
  int rowA = partsA[1].toInt();
  int rowB = partsB[1].toInt();
  if (rowA != rowB)
    return rowA < rowB;

  // Compare Seat (Numeric)
  return partsA[2].toInt() < partsB[2].toInt();
}

void PdfsToTxt::runCleaner() {
  // Security Check
  SecurityManager::instance().checkAndAct();

  QString folderPath = m_pathEdit->text();
  if (folderPath == "No folder selected..." || !QDir(folderPath).exists()) {
    QMessageBox::warning(this, "Warning",
                         "Please select a valid PDF folder first.");
    return;
  }

  log("Initializing process...", true);
  QStringList cleanNames;

  // Scan Engine
  QDirIterator it(folderPath, QStringList() << "*.pdf", QDir::Files,
                  QDirIterator::Subdirectories);
  QRegularExpression nameRegex("^([A-Za-z0-9]+-\\d+-\\d+)");

  while (it.hasNext()) {
    it.next();
    QString fileName = it.fileInfo().baseName();
    auto match = nameRegex.match(fileName);
    if (match.hasMatch()) {
      cleanNames.append(match.captured(1));
    }
  }

  if (cleanNames.isEmpty()) {
    log("❌ ERROR: No matching PDF names found (pattern: S-R-S).");
    return;
  }

  // Natural sort
  std::sort(cleanNames.begin(), cleanNames.end(), PdfsToTxt::naturalSort);

  // Save on Desktop
  QString desktopPath =
      QStandardPaths::writableLocation(QStandardPaths::DesktopLocation);
  QString outputFile = desktopPath + "/Clean_PDF_List.txt";

  QFile file(outputFile);
  if (file.open(QIODevice::WriteOnly | QIODevice::Text)) {
    QTextStream out(&file);
    for (const QString &name : cleanNames) {
      out << name << "\n";
    }
    file.close();

    log(QString("SUCCESS: %1 PDFs processed.").arg(cleanNames.size()));
    log(QString("Saved to Desktop: Clean_PDF_List.txt"));

    log("\nPreview (First 10):");
    for (int i = 0; i < std::min(10, (int)cleanNames.size()); ++i) {
      log(QString("   %1. %2").arg(i + 1, 2).arg(cleanNames[i]));
    }

    QMessageBox::information(
        this, "Done",
        QString("List generated successfully!\nTotal: %1\nCheck your Desktop.")
            .arg(cleanNames.size()));

    // Auto-open
    QDesktopServices::openUrl(QUrl::fromLocalFile(outputFile));
  } else {
    log("❌ ERROR: Could not save file.");
  }
}

} // namespace GOL

File: legacy_python/scripts/pdfs_to_TXT.py
import os
import re
import customtkinter as ctk
import requests
import subprocess
import sys
from tkinter import messagebox, filedialog

# ================== [ SECURITY & DESIGN CONFIG ] ==================
VERSION = "2.9"
FIREBASE_URL = "https://golevents1-default-rtdb.europe-west1.firebasedatabase.app"
ACCENT_COLOR = "#1f6aa5" 
BG_COLOR = "#0b0b0b"
CARD_BG = "#161616"
CARD_BORDER = "#252525"
# =================================================================

def get_hwid():
    try:
        cmd = 'wmic baseboard get serialnumber'
        hwid = str(subprocess.check_output(cmd, shell=True), 'utf-8').split('\n')[1].strip()
        return hwid
    except:
        import hashlib
        return hashlib.sha256(os.environ['COMPUTERNAME'].encode()).hexdigest()[:16]

# --- SECURITY WRAPPER ---
class SecurityCheck(ctk.CTk):
    def __init__(self, key):
        super().__init__()
        self.withdraw()
        self.license_key = key
        
        self.splash = ctk.CTkToplevel(self)
        self.splash.title("Authenticating Module...")
        self.splash.geometry("450x250")
        self.splash.configure(fg_color=BG_COLOR)
        self.splash.overrideredirect(True)
        
        x = (self.winfo_screenwidth() // 2) - 225
        y = (self.winfo_screenheight() // 2) - 125
        self.splash.geometry(f"+{x}+{y}")

        ctk.CTkLabel(self.splash, text="GOLEVENTS SECURE 👑", font=("Arial", 24, "bold"), text_color=ACCENT_COLOR).pack(pady=(45, 5))
        self.status_lbl = ctk.CTkLabel(self.splash, text="Verifying Module Access...", font=("Arial", 13), text_color="gray")
        self.status_lbl.pack()
        
        self.progress = ctk.CTkProgressBar(self.splash, width=350, height=12, progress_color=ACCENT_COLOR)
        self.progress.set(0)
        self.progress.pack(pady=30)
        
        self.after(500, self.verify)

    def verify(self):
        hwid = get_hwid()
        try:
            res = requests.get(f"{FIREBASE_URL}/keys/{self.license_key}.json").json()
            if res and res.get("status") == "active" and res.get("hwid") == hwid:
                self.progress.set(1.0)
                self.status_lbl.configure(text="ACCESS GRANTED", text_color="#2ecc71")
                self.after(800, self.launch_main_tool)
            else:
                messagebox.showerror("Security Alert", "Unauthorized Access!\nPlease launch through GOLEVENTS Hub.")
                sys.exit()
        except:
            sys.exit()

    def launch_main_tool(self):
        self.splash.destroy()
        self.deiconify()
        PDFNameCleanerApp(self)

# --- NATURAL SORT LOGIC ---
def natural_sort(key):
    try:
        p1, p2, p3 = key.split('-')
        if p1.isdigit():
            p1 = (0, int(p1))
        else:
            p1 = (1, p1)
        return (p1, int(p2), int(p3))
    except:
        return ((9, ""), 9999, 9999)

# --- GUI INTERFACE ---
class PDFNameCleanerApp:
    def __init__(self, master):
        self.master = master
        self.master.title("GOLEVENTS - PDF NAME CLEANER PRO 📄")
        self.master.geometry("800x700")
        self.master.configure(fg_color=BG_COLOR)

        # Header Section
        header_frame = ctk.CTkFrame(self.master, fg_color="transparent")
        header_frame.pack(fill="x", padx=40, pady=(30, 20))
        
        ctk.CTkLabel(header_frame, text="📄 PDF NAME CLEANER", font=("Arial", 30, "bold"), text_color=ACCENT_COLOR).pack(anchor="w")
        ctk.CTkLabel(header_frame, text="Extract 'Sector-Row-Seat' only from filenames into a TXT list.", font=("Arial", 13), text_color="gray60").pack(anchor="w")

        # Action Frame
        action_frame = ctk.CTkFrame(self.master, fg_color=CARD_BG, corner_radius=15, border_width=1, border_color=CARD_BORDER)
        action_frame.pack(fill="x", padx=40, pady=10)

        self.path_var = ctk.StringVar(value="No folder selected...")
        ctk.CTkEntry(action_frame, textvariable=self.path_var, font=("Arial", 12), width=500, height=45, 
                      fg_color="#0d0d0d", state="readonly").pack(side="left", padx=20, pady=20)
        
        ctk.CTkButton(action_frame, text="BROWSE", width=120, height=45, fg_color=ACCENT_COLOR, 
                      font=("Arial", 13, "bold"), command=self.browse_folder).pack(side="left", padx=(0, 20))

        # Process Button
        self.process_btn = ctk.CTkButton(self.master, text="🚀 START EXTRACTION & SAVE", height=55, 
                                          fg_color="#10b981", hover_color="#059669", text_color="black",
                                          font=("Arial", 16, "bold"), command=self.run_cleaner)
        self.process_btn.pack(pady=20, padx=40, fill="x")

        # Log Area
        self.log_area = ctk.CTkTextbox(self.master, font=("Consolas", 12), fg_color="#050505", 
                                        text_color="#00FF94", border_width=1, border_color=CARD_BORDER)
        self.log_area.pack(fill="both", expand=True, padx=40, pady=(0, 40))

    def browse_folder(self):
        p = filedialog.askdirectory()
        if p:
            self.path_var.set(p)

    def log(self, msg, clear=False):
        self.log_area.configure(state="normal")
        if clear: self.log_area.delete("1.0", "end")
        self.log_area.insert("end", f"> {msg}\n")
        self.log_area.see("end")
        self.log_area.configure(state="disabled")

    def run_cleaner(self):
        folder_path = self.path_var.get()
        if folder_path == "No folder selected..." or not os.path.isdir(folder_path):
            messagebox.showwarning("Warning", "Please select a valid PDF folder first.")
            return

        self.log("Initializing process...", clear=True)
        clean_names = []

        # Logic Engine
        for root, dirs, files in os.walk(folder_path):
            for filename in files:
                if filename.lower().endswith(".pdf"):
                    name = filename.rsplit('.', 1)[0]
                    # pattern: sector-row-seat
                    match = re.match(r'^([A-Za-z0-9]+-\d+-\d+)', name)
                    if match:
                        clean_names.append(match.group(1))

        if not clean_names:
            self.log("❌ ERROR: No matching PDF names found (pattern: S-R-S).")
            return

        # Natural sort
        clean_names.sort(key=natural_sort)

        # Save TXT on Desktop
        desktop = os.path.join(os.path.expanduser("~"), "Desktop")
        output_file = os.path.join(desktop, "Clean_PDF_List.txt")

        try:
            with open(output_file, "w", encoding="utf-8") as f:
                for name in clean_names:
                    f.write(name + "\n")
            
            self.log(f"SUCCESS: {len(clean_names)} PDFs processed.")
            self.log(f"Saved to Desktop: Clean_PDF_List.txt")
            
            # Preview first 10
            self.log("\nPreview (First 10):")
            for i, name in enumerate(clean_names[:10]):
                self.log(f"   {i+1:2d}. {name}")
            
            messagebox.showinfo("Done", f"List generated successfully!\nTotal: {len(clean_names)}\nCheck your Desktop.")
            
            # Auto-open
            os.startfile(output_file)
        except Exception as e:
            self.log(f"❌ ERROR: Could not save file. {e}")

# --- MAIN EXECUTION ---
if __name__ == "__main__":
    if len(sys.argv) > 1:
        app = SecurityCheck(sys.argv[1])
        app.mainloop()
    else:
        ctk.set_appearance_mode("Dark")
        root = ctk.CTk()
        root.withdraw()
        messagebox.showerror("Unauthorized", "Please launch via GOLEVENTS Hub.")
        sys.exit()